[Consensus] EpochManager as a single SoT for epoch information

gterzian opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Libra project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

FIX #1671

 And also make a first step towards https://github.com/libra/libra/issues/2152

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

gterzian commented:
@davidiw I wanted to ask for some early feedback on the proposed re-design of the network task/consensus container stuff. 

At this point, I only know that the `consensus` crate actually compiles with these changes. I haven't tried to compiler the entire project, or run it, yet. 

I was thinking, before I even begin to look at adapting the current unit-tests, I better get some feedback on the design.

Couple of highlights:

1. The `NetworkTask` doesn't own any `EpochInfo` anymore.
2. I removed all use of `libra_channel`, instead the `NetworkTask` itself is used a a kind of buffer/channel for incoming messages. 
3. I tried to build a more coherent processing pipeline, like:
    - `NetworkTask` handles incoming `ConsensusMsg`, buffers/drops them locally as `FromNetworkMsg`. 
    - `EpochManager` handles those `FromNetworkMsg`, focusing on the epoch logic. If necessary, it then makes a `EpochMsg` available for further processing.
    - `EventProcessor` processes available `EpochMsg` for the current epoch. 

Let me know what you think!

davidiw commented:
@gterzian thanks for this PR!

Could you do me a couple of favors? Make sure you pass tests locally before requesting review, so I focus more on readability rather than correctness. Merge your commits as you move along especially since I haven't reviewed anything yet.

With that in mind, I wonder if we're trying to tackle too large a problem initially. My first question would be is do we benefit from having this separate network stack. My suspicion is not really and if we do, we could always figure a way to integrate that more cleanly. Also we have intention to remove the tokio layer from within consensus and need to build a need timer to handle that. The biggest contention is that we have a network layer doing pre-processing work. This is somewhat of an optimization that may not actually be that usable considering that the current deployment model will only have trusted network connections and that networking employs round robin.

A couple of proposals to think about:
* Make a single select loop with a single thread / worker, we have internal tools evaluate performance if you can put together a commit for this
* Have the network loop take context from epoch manager about changes in Epochs (i.e., have a channel from epoch manager to the outside world, where epoch manager can update about epoch changes).

The former makes the code substantially simpler but risks causing some perf regression. The latter is probably reasonable no matter what. Personally, I would probably do lighter weight cleanups first and then tackle more complex, riskier cleanups. This definitely fits more in the risky, whereas stripping out tokio isn't so much.

gterzian commented:
> This definitely fits more in the risky, whereas stripping out tokio isn't so much.

Interesting, I was thinking the opposite :)

I would be in favor of following the "select in a single thread" approach, it's probably how I would have done this from scratch. 

However, it's a fact that you are currently using Tokio, and already have in place what you refer to as a "network stack". 

I agree this additional network worker used for buffering is not necessarily useful, and the thing  is that is what you currently have. 

So what I did was actually trying to not change things too much, while fixing the incoherentness(as I perceived it), where the network stack would actually handle some epoch logic. 

I think trying to remove Tokio would actually be more complicated and risky, since you have async API both at the stream of incoming events, and at the outgoing network messages(See [`NetworkSender`](https://github.com/libra/libra/blob/36dfbdbc7b4172210697486de9cc0e68ee901dae/consensus/src/chained_bft/network.rs#L82)).

So, I was actually thinking restructuring the processing flow, as this PR does, while leaving the async stuff in place, might be the easier first step. 

> Also we have intention to remove the tokio layer from within consensus and need to build a need timer to handle that.

As a side note, if using a thread, you can quite easily build a timer from basic utilities like https://docs.rs/crossbeam-channel/0.4.0/crossbeam_channel/fn.after.html

> Make sure you pass tests locally before requesting review, so I focus more on readability rather than correctness.

I'd be happy to do that. The unit-test appear even more complicated then the code itself. There is no point trying to change those without first knowing if the design is appropriate or not.

gterzian commented:
@davidiw One last thing I would like to add to https://github.com/libra/libra/pull/2402#issuecomment-581733340

In some way, the proposed re-design here would result in a structure that is somewhat agnostic of it's runtime context, hence easier to eventually refactor to a native thread if necessary. 

For example, I've moved all business logic out of the channel(`libra_channel`), and into data owned by the task. So it happens to still use an async channel, but it's sort of agnostic to the "asyncness" of the channel. You could literally swap it out with a crossbeam channel and run it in a thread and the overall logic would stay the same. 

It would be much harder to adapt `libra_channel` to run in a thread(using a `Condvar` as the  waker and so on, also not sure you would want to do that at all, it would be better to simply select directly on a set of crossbeam channels and have the biz logic be local to a thread).

davidiw commented on consensus/src/chained_bft/network.rs:
```diff
@@ -473,15 +639,8 @@ impl<T: Payload> NetworkTask<T> {
         peer_id: AccountAddress,
     ) -> anyhow::Result<()> {
         let sync_info = SyncInfo::try_from(sync_info)?;
-        match sync_info.epoch().cmp(&self.epoch()) {
```
Perhaps we could start with a smaller commit that just unifies the common logic on reviewing the messages epoch against the current epoch? There looks like a lot of copied code that could be isolated into a common function:

`
  fn check_epoch(&self, epoch: u64) {
      ...
   }
`

gterzian replied:
Yes this additional change could be made in a separate commit indeed.

gterzian commented:
> The challenge I'm having is that this is a pretty large PR. The overall concept clicks but I don't have the cycles to parse the complexity. 

Ok, let's see if we can make it easier to review in some way then.

> I think if we can break down the code into smaller commits that address specific concerns, it would be more digetable.

So I think this might be a bit hard to do, because the changes are quite dependent on each other. 

For example, it would be nice to have a commit moving the epoch logic from the `NetworkTask` to the `EpochManager`, and that actually requires re-structuring both event-loops as I have done, since that precludes the continued use of separate channels by the `NetworkTask` with names like `different_epoch`.

How about I tried to write down a bit more clearly what I've done in English, with links to the code?

> One question that comes to mind in both looking at your comments and your code is why we're switching to a different channel.

> Also fwiw, the SoT for epoch information was obtained by leveraging the Arc<RwLock, whereas before they used a channel in this code, which could result in some weirdness if two internal events were racing against each other.

Ok so, the goal is really to move all epoch logic to the `EpochManager`, and away from the `NetworkTask`. 

Why? Because it would enable the `EpochManager` to manage all epoch logic robustly. 

By the way, this goes a bit further then making the  `EpochManager` a SoT, because a SoT sort of implies you're still sharing this truth with other concurrent components. So the weirdness you mention when a channel was previously used, I'm pretty sure that will not be the case when you switch from a SoT model to a non-shared, "all epoch info is owned by a single component", approach. 

Although the issue linked to this PR mentions a potential race, after having worked on this, I actually think the issue had already been fixed with the `Arc<RwLock<EpochInfo>>`. 

However, it's one thing to have fixed a specific race, it's entirely another to never have to deal with these kind of problems ever again. The current re-design is a step in that direction, in my opinion. 

Likewise, the "queueing", "buffering", and "clearing" of messages when the epoch changes, is also something that can be made more robust by moving this logic, and the related data, entirely into one concurrent component(be it a task, or a thread). 

So I've done that as well, by removing `libra_channel`, and moving the related business logic entirely to the `NetworkTask`. 

The result is that the `NetworkTask` now manages some local maps and queues, with aren't themselves concurrent or shared with the `EpochManager`. That makes the queuing logic entirely single-threaded. 

Then, since there is still a need to communicate between the `NetworkTask` and the `EpochManager`, I switched to using two unbounded channels. Why unbounded? Because then you never have to worry about deadlocking those two components. How about running out of memory? No problem, the bounds of the overall "work queue" is expressed in the local logic and data of the `NetworkTask`, there is no need to further bound the actual communication channel. 

Finally, to actually allow the `NetworkTask` to manage the buffering effectively, I had to switch to a "pull" workflow, where the `EpochManager` requests data from the `NetworkTask` when it needs it. Why? Because otherwise, I would have to implement a concurrent queue managing the buffering, essentially what `libra_channel` was doing. The goal is moving the queuing logic to be local to the `NetworkTask`. Then we cannot have the `NetworkTask` just push data, since then there would be effectively no buffering. So instead the `EpochManager` can pull it. 

Also note that it is also the `EpochManager` that "tells" the `NetworkTask` to clear it's data when the epoch changes. This is done by sending a message on the same channel on which data is requested. And actually when an epoch changes, the `EpochManager` simply sends two messages, one to clear the data, another to ask for more. The side-effect of handling these two messages will be local to the `NetworkTask`, and guaranteed to happen in the order that they were sent by the `EpochManager`.

So essentially, the point of this PR is to move two pieces of logic: 1. epoch management, and 2. messages buffering, dropping, and clearing, away from a shared-state model, in into two different component each owning local data.

Even though it might be that the lock currently used around `EpochInfo` was effective in fixing a specific race, the overall structure of the current code leaves what is essentially a hornet's nest of concurrency issue waiting to be kicked over. 

A simple example to avoid the above sounding like some empty statement:

This is the current epoch change code:

```rust
// Start critical section on epoch_manager.epoch_info,
// because start_new_epoch will acquire the write lock and set the new epoch.

event_processor = epoch_manager.start_new_epoch(ledger_info).await;

// End of the critical section on epoch_manager.epoch_info.

// WARNING: we're not holding the lock on epoch_manager.epoch_info!
// Yet we expect all messages of the old epoch to be cleared, 
// and also we expect all messages from the old epoch to from now on to be enqueued to 
// the `different_epoch` channel. 
//
// clean up all the previous messages from the old epochs
network_receivers.clear_prev_epoch_msgs();

event_processor.start().await;
```

So, basically, the `Rwlock<EpochInfo>` only protects that specific state. It doesn't give you any guarantee about the internals of `network_receivers`. It also doesn't give you any guarantee about scheduling of tasks, or compiler optimizations. For example, since you're using a threaded scheduler, the `NetworkTask` is probably running in parallel to this code. It's unclear what can happen in terms of new messages being enqueued in the various sub-channels of `network_receivers` while this epoch changes occurs.

One should either pair the `Rwlock<EpochInfo>` with a `Condvar`, and write some robust logic both in `NetworkTask`, and in the consensus loop, to ensure you never get anything weird. 

Or, my preferred alternative, you isolate state into one task(or thread). Then you only write logic using local data, and communicate via an unbounded queue with the other parallel component, ensuring your logic doesn't make any assumptions about when the message will be handled. 

One thing I would avoid, is using channels **and** shared-state, and assume the two are going to synchronize implicitly. 

----

I'm note sure if this makes it less complex to review. I do think it raises some important point about the actual need for a re-design, and hopefully brings further clarity to the proposal.

gterzian commented:
> if we can make it so that we either don't need to modify integration tests or do as much as possible before we do have to modify them, that would make the process substantially smoother

I will take another look at those, and I'm actually somewhat hopeful the changes could be minimal, since we're essentially just swapping `network_receiver` for another channel.

bors-libra commented:
:umbrella: The latest upstream changes (presumably f84e7f0fe6998fa227078d793c310ef44ad47f84) made this pull request unmergeable. Please resolve the merge conflicts.

davidiw commented on consensus/src/chained_bft/chained_bft_smr.rs:
```diff
@@ -88,61 +90,98 @@ impl<T: Payload> ChainedBftSMR<T> {
+                            Ok(EpochCheck::NotCurrent) => {
+                                // No further processing required of an EpochMsg,
+                                // so we just request more data and go back to the select above.
+                                let _ = data_request_sender.unbounded_send(ConsensusDataRequest::NewData);
```
It isn't intuitive to me that we are asking the network for data.

gterzian replied:
Why not?

gterzian commented on consensus/src/chained_bft/epoch_manager.rs:
```diff
@@ -88,7 +108,101 @@ impl<T: Payload> EpochManager<T> {
+    ) -> anyhow::Result<EpochCheck<T>> {
+        match msg {
+            FromNetworkMsg::Proposal(proposal) => {
+                if proposal.epoch() != self.epoch_info.epoch {
```
I see how it could make sense to "unify this repetitive code", and I don't see how it would somehow reduce the complexity of the rest of this PR. This code was previously in `NetworkTask`. 

The commit reducing the repetition could be completely separate from this PR, and done in a follow-up. How does doing it first make the overall restructuring easier?

gterzian replied:
Ok I separated this epoch checking into a separate method. Every little bit helps!

gterzian commented on consensus/src/chained_bft/epoch_manager.rs:
```diff
@@ -282,6 +396,7 @@ impl<T: Payload> EpochManager<T> {
             validators,
-        )
+        );
+        Box::new(event_processor)
```
Clippy was complaining about different sizes of the enum variants.

gterzian commented on consensus/src/chained_bft/network.rs:
```diff
@@ -274,142 +252,357 @@ impl NetworkSender {
+                message = all_events.next() => {
+                    match message {
+                        Some(Ok(msg)) => self.handle_msg(msg).await,
+                        _ => continue,
```
I'm actually not sure, could be that the stream has ended? It might actually always be `Ok` here. 

There is a `complete` handler below, which I believe is called when the select is "completed". 

In terms of exiting, my understanding is that when the `ConsensusProvider` drops the runtime, all tasks are cancelled. This happens when `stop` is called.

gterzian replied:
Switched to using `select_next_some` on this one...

gterzian commented:
@davidiw Are you already regretting suggesting I work on this issue? :)

More seriously, I understand you concern. Believe me, my goal is not introducing more complexity. And I simply have a different view than you on this. 

As I have stated earlier, your suggestion are something I agree with(using threads and so on), and I don't see how removing Tokio, using threads, and the other things you propose, would in any way be simpler than the current changes. I think they would be bigger. The current changes at least do not change the basic execution model of those components, leaving it based on Tokio. 

> The entire intent of this PR is to improve accessibility, but it increases the number of lines by over 350.

I think counting the lines here is a misguided approach. I removed the use of `libra_channel`, which included shared-state concurrent code. I've replaced this with new code that is single-threaded. Those do not have the same weight in terms of complexity, even thought they add line numbers to the code. 

Finally, the reason why I do all this in one PR, and probably one commit, is simply because the current is completely interdependent. To solve this requires a "big" restructuring. I don't see how this can be done in independent steps, and that's a result of the current code. The logic is simply intertwined between `NetworkTask` and the other concerns. 

This PR introduces clearer separation. This initial restructuring might be a bitter pill to swallow, yet it's necessary. Follow-up work, including doing the stuff you suggest with threads and so on, would be much easier.

gterzian commented:
Closing dormant PR.

davidiw commented:
> Closing dormant PR. It seems the code is moving in a different direction, and we also have a different view on where to take this. @davidiw Thanks for the comments so far.

Really appreciate the attempt. I would be game for working with you again but perhaps first spending more time brainstorming first and aligning on a solution so that it makes it easier to ensure the work can land :).
